src/resolver.rs

```diff
@@ -12,6 +12,20 @@ pub fn resolve<D: Dialect>(ast: &mut Block) -> BTreeMap<u64, FunctionSignature>
     std::mem::take(&mut r.function_signatures)
 }
 
+/// Resolves all references in `to_resolve` given an already resolved ast `reference`.
+/// This can be used to resolve e.g. an expression to be evaluated.
+/// This only considers symbols defined at the top level of the reference block.
+pub fn resolve_inside<D: Dialect>(to_resolve: &mut Expression, reference: &Block) {
+    let mut r = Resolver::<D>::new();
+    r.enter_block_immut(reference);
+    for statement in &reference.statements {
+        if let Statement::VariableDeclaration(var) = &statement {
+            r.exit_variable_declaration_immut(var);
```

Comment:
why?

Reply:
I split the enter/exit functions into versions that take an immutable reference. It seems a trait that takes a mutable reference cannot be implemented by a function taking an immutable reference. I did it because I wanted `reference` to be immutable.